File: base-ui/src/main/java/com/adyen/checkout/base/ui/view/AdyenTextInputEditText.java
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 26/8/2019.
 */

package com.adyen.checkout.base.ui.view;

import android.content.Context;
import android.text.Editable;
import android.text.InputFilter;
import android.text.TextWatcher;
import android.util.AttributeSet;

import androidx.annotation.CallSuper;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.material.textfield.TextInputEditText;

public class AdyenTextInputEditText extends TextInputEditText {

    protected static final int NO_MAX_LENGTH = -1;

    @Nullable
    private Listener mListener;

    public AdyenTextInputEditText(@NonNull Context context) {
        this(context, null);
    }

    public AdyenTextInputEditText(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    /**
     * Constructor of AdyenTextInputEditText.
     */
    public AdyenTextInputEditText(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr == 0 ? com.google.android.material.R.attr.editTextStyle : defStyleAttr);
        addTextChangedListener(getTextWatcher());
    }

    public void setOnChangeListener(@NonNull Listener listener) {
        this.mListener = listener;
    }

    @NonNull
    public String getRawValue() {
        return getText() != null ? getText().toString() : "";
    }

    @CallSuper
    protected void afterTextChanged(@NonNull Editable editable) {
        if (mListener != null) {
            mListener.onTextChanged(editable);
        }
    }

    protected void enforceMaxInputLength(int maxLength) {
        if (maxLength != NO_MAX_LENGTH) {
            setFilters(new InputFilter[]{new InputFilter.LengthFilter(maxLength)});
        }
    }

    @NonNull
    private TextWatcher getTextWatcher() {
        return new TextWatcher() {

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                // Empty
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                // Empty
            }

            @Override
            public void afterTextChanged(Editable s) {
                AdyenTextInputEditText.this.afterTextChanged(s);
            }
        };
    }

    public interface Listener {
        void onTextChanged(@NonNull Editable editable);
    }
}

File: ui-core/src/main/java/com/adyen/checkout/components/ui/view/AdyenTextInputEditText.kt
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 26/8/2019.
 */
package com.adyen.checkout.components.ui.view

import android.content.Context
import android.text.Editable
import android.text.InputFilter
import android.text.InputFilter.LengthFilter
import android.text.TextWatcher
import android.util.AttributeSet
import androidx.annotation.CallSuper
import com.google.android.material.R
import com.google.android.material.textfield.TextInputEditText

open class AdyenTextInputEditText @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : TextInputEditText(context, attrs, if (defStyleAttr == 0) R.attr.editTextStyle else defStyleAttr) {

    private var listener: Listener? = null

    open val rawValue: String
        get() = text?.toString() ?: ""

    /**
     * Constructor of AdyenTextInputEditText.
     */
    init {
        this.addTextChangedListener(textWatcher)
    }

    fun setOnChangeListener(listener: Listener) {
        this.listener = listener
    }

    @CallSuper
    protected open fun afterTextChanged(editable: Editable) {
        // TODO temporary measure to avoid notifying changes for disabled edit texts.
        //  This should be removed after the view model starts emitting the final formatted strings for the edit texts
        if (isEnabled) listener?.onTextChanged(editable)
    }

    protected fun enforceMaxInputLength(maxLength: Int) {
        if (maxLength != NO_MAX_LENGTH) {
            filters = arrayOf<InputFilter>(LengthFilter(maxLength))
        }
    }

    private val textWatcher: TextWatcher
        get() = object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) = Unit

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) = Unit

            override fun afterTextChanged(s: Editable) {
                this@AdyenTextInputEditText.afterTextChanged(s)
            }
        }

    fun interface Listener {
        fun onTextChanged(editable: Editable)
    }

    companion object {
        protected const val NO_MAX_LENGTH = -1
    }
}
